components/layout_2020/display_list/stacking_context.rs

```diff
@@ -464,22 +470,24 @@ impl BoxFragment {
     ) {
         // If we are creating a stacking context, we may also need to create a reference
         // frame first.
-        let relative_border_rect = self
-            .border_rect()
-            .to_physical(self.style.writing_mode, &containing_block_info.rect);
-        let border_rect =
-            relative_border_rect.translate(containing_block_info.rect.origin.to_vector());
-        let established_reference_frame =
-            self.build_reference_frame_if_necessary(builder, &border_rect);
+        let reference_frame_data = self.reference_frame_data_if_necessary(containing_block_info);
 
         // WebRender reference frames establish a new coordinate system at their origin
         // (the border box of the fragment). We need to ensure that any coordinates we
         // give to WebRender in this reference frame are relative to the fragment border
         // box. We do this by adjusting the containing block origin.
         let mut new_containing_block_info = containing_block_info.clone();
-        if established_reference_frame {
-            new_containing_block_info.rect.origin =
-                (-relative_border_rect.origin.to_vector()).to_point();
+
+        if let Some(reference_frame_data) = &reference_frame_data {
```

Comment:
Perhaps this could all be part of a method called `build_reference_frame_if_necessary(containing_block_info, new_containing_block_info)`. It might be good to isolate all of the complications of setting up the ReferenceFrame into one method.